### DISCUSSION TITLE
Problem with bounds detection, Bounds is alway Vec3.Zero

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#1014] [Q&A]
@maluoi Nick, I'm sure I don't understand the documentation. As long as I have been developed using SK, I still can't seem to fathom the bounds and collisions. I grabbed a snippet from one of your demo files and I simply want to detect when my fingertip is in the bounds of a model and I want a collection of objects to be moveable. That should be dead simple. But it apparently is not. It seems that Bounds always returns vec zero as the center. That makes no sense at all. All it means is the model's origin is zero. Also, if I move the model using the grab, the Pose clearly updates its position and rotation but the bounds does not. Whatever rotation and position was initially set remains fixed in space, bounds does not continue to "wrap" around the model. Finally, it seems that since the bounds.center is always zero the hand only will detect/collide at its position zero. Having my hand position colliding with a model wherever it is in space and whatever its orientation seems like it should be a no-brainer.

How do I collide with a bounded object that moves in a UI.Handle? The following code attempts to have the fingertip collide with the bounds of the model. In fact, the bounds collision space is centered on vec3 zero, NOT the model. If I move the model the model bounds do not move with the model and remain fixed in the original location and original rotation. My understanding is the UI.Handle should manage the hierarchy transforms including rotation? I would really appreciate some assistance in understanding what is really happening.

```
public void Step()
{
	if (SK.System.displayType == Display.Opaque)
		Default.MeshCube.Draw(floorMaterial, floorTransform);

	// pipe sphere and Ray
	bool pipeRayActive = UI.HandleBegin("PipeRay", ref posePipeRay, model.Bounds, false, UIMove.Exact);
	model.Draw(Matrix.S(scale), pipeColor);

	Sphere sphere = new Sphere(spherePos, 0.0625f);
	Vec3 sphereDir = Vec3.AngleXZ(Time.Totalf * 90, SKMath.Cos(Time.Totalf * 3) * 0.375f + 0.05f).Normalized;
	Ray sphereRay = new Ray(sphere.center - sphereDir * 0.0875f, sphereDir);

	Lines.Add(sphereRay.position, sphereRay.At(0.05f), colTest, 2 * Units.mm2m);
	if (sphereRay.Intersect(sphere, out Vec3 sphereAt))
		Mesh.Sphere.Draw(Material.Default, Matrix.TS(sphereAt, 0.005f), colIntersect);
	Mesh.Sphere.Draw(Material.Default, Matrix.TS(sphere.center, 0.0625f), colObj);


	Hand hand = Input.Hand(1);
	Pose fingertip = hand[FingerId.Index, JointId.Tip].Pose;
	if (model.Bounds.Contains(fingertip.position))
		spherePos.x = fingertip.position.x;

	UI.HandleEnd();
}

```

**Comment by mvvvv** [comment]
Have a look to the examples :
https://stereokit.net/Pages/StereoKit/Bounds.html
https://stereokit.net/Pages/StereoKit/Ray/Intersect.html

// Note that positions must be in a coordinate space relative to  the bounds!

**Reply by maluoi** [reply]
Yep! This is it :) Model.Bounds is in model space, while fingertip.position is in world space.

**Comment by laultman** [comment]
@maluoi This answers the question about the documentation at least. Since the bounds didn't really achieve what I wanted anyway, I went a different route to get closer to my ultimate goal.
I have some code (below) that base very loosely on the RayMesh demo. In my case, I have a pipe (cylinder) that I want to determine the position of my hand relative to a point along the pipe. Mesh intersect does not work because it only applied to the origin point. The problem was again the local and world space coordinates. I used a matrix transform to get the ray into the local mesh space. The hand acts on the "probe" sphere that is tracking the position of the target sphere.
This works great except for one thing. When the pipe is moved it is rotated. The result is the sphere no longer tracks along the pipe. When the pipe and the sphere are in the same space i.e. UI.HandleBegin ... End the rotation is handled by the hierarchy transform.
How would you rotate the sphere's scanPose transform?
```
		public void ObjectDetectorDraw()
		{
			UI.Handle("PipeRay", ref posePipeRay, pipeMesh.Bounds, false, UIMove.Exact);
			pipeMesh.Draw(pipeMat, posePipeRay.ToMatrix());
			scanPose= posePipeRay;

			spherePos = scanPose.position;
			spherePosX = ScanPipeX(spherePosX, pipeScanIncrement);

			// Create a ray that's in the pipe mesh's model space
			Matrix transform = scanPose.ToMatrix();
			var t = transform.Translation;
			t.x = spherePosX;
			Sphere sphere = new Sphere(t, 0.0625f);

			Ray ray = transform
				.Inverse
				.Transform(Ray.FromTo(castPose.position, t));
			Lines.Add(castPose.position, t, Color.White, 0.005f);
			pipeSphereMesh.Draw(sphereMat, Matrix.TS(sphere.center, 0.0625f));       // pipe scan sphere


			UI.Handle("Cast", ref castPose, castSphereMesh.Bounds * 0.03f);
			castSphereMesh.Draw(Default.MaterialUI, castPose.ToMatrix(.035f));


			UI.WindowBegin("Radiation Frisk", ref DataWindowPose, V.XY(0.25f, 0), UIWin.Normal, UIMove.FaceUser);
			var distCM = Vec3.Distance(t, castPose.position) * metersScale; // scale meters by factor
			var mradRate = PHNS.CalculateScanDoseRate_mRhr(microCi, distCM, 0.001f).ToString("n4");
			var pradRate = PHNS.CalculateScanDoseRate_mRhr(picoCi, distCM, 0.001f).ToString("n5");

			Vec2 labelSize = V.XY(0.08f, 0);
			UI.Label("Zone", labelSize); UI.SameLine();
			UI.Label(Vec3ToString(t), false);
			UI.Label("Probe", labelSize); UI.SameLine();
			UI.Label(Vec3ToString(castPose.position), false);
			UI.Label("Distance", labelSize); UI.SameLine();
			UI.Label($"{distCM.ToString("n2")} cm", false);
			UI.Label("Radiation", labelSize);
			UI.Label($"microCi: {mradRate}", false); UI.SameLine();
			UI.Label($" picoCi: {pradRate}", false);

			UI.HSeparator();
			UI.Text("Raidation levels are calculated using two different Ci scales, micro and pico. The distance from the probe is measured in centimeters.");
			UI.HSeparator();
			if (UI.Button("Close"))
				SK.Quit();

			UI.WindowEnd();
		}
		private float ScanPipeX(float x, float increment)
		{
			x += increment;
			if (x >= halfPipe)
				x = -EngineeringData.totalPipeLength / 2;
			return x;
		}

		private string Vec3ToString(Vec3 v)
		{
			return $"{v.x.ToString("n2")}, {v.y.ToString("n2")}, {v.z.ToString("n2")}";
		}

```

**Comment by laultman** [comment]
@maluoi Speaking of colliders and solid bodies. With the demise of the physics lib in SK. There are many requirements for simple colliders that are very useful but don't necessarily require a full physics lib. HoloLens for example, the bulk of its use cases would not need object physics as it is not really a game device nor is it particularly good at model simulation. However, in the real-world HoloLens goes well beyond today's definition of mixed reality, having mesh and object collision detection is very useful. Mainly in the UI where users are interacting with real physical objects and holographic objects as part of the job operation.
I found an [article](https://developer.mozilla.org/en-US/docs/Games/Techniques/3D_collision_detection) that explains rudimentary collision detection that is simple to implement. These are not pixel quality colliders, but for UI purposes are adequate. I just needed a few for a project, but it is conceivable that all the built in primitives could have collider extensions.
I offer this code for possible inclusion and hopefully others will enhance it further. Where I was hoping you would work with your skill in C++ would be integrate mesh collision along the line of ray casting to a mesh face.
```
public static class ColliderExtensions
{
	public static bool SpheresCollision(this Sphere sphereA, Sphere sphereB)
	{
		return Vec3.Distance(sphereA.center, sphereB.center) < sphereA.radius + sphereB.radius;
	}

	public static bool SphereBoxCollision(this Sphere sphere, Vec3 spherePosition, Bounds box, Vec3 boxPosition)
	{
		var half = box.dimensions * 0.5f;
		var min = boxPosition - half;
		var max = boxPosition + half;

		var closestPoint = new Vec3(
			Math.Max(min.x, Math.Min(spherePosition.x, max.x)),
			Math.Max(min.y, Math.Min(spherePosition.y, max.y)),
			Math.Max(min.z, Math.Min(spherePosition.z, max.z))
		);

		return Vec3.Distance(closestPoint, spherePosition) < sphere.radius;
	}

	public static bool IsPointInsideSphere(this Sphere sphere, Vec3 point)
	{
		return Vec3.Distance(sphere.center, point) < sphere.radius;
	}

	public static bool IsPointInsideBox(this Bounds box, Vec3 boxPosition, Vec3 point)
	{
		var half = box.dimensions * 0.5f;
		var min = boxPosition - half;
		var max = boxPosition + half;

		return point.x >= min.x && point.x <= max.x &&
				point.y >= min.y && point.y <= max.y &&
				point.z >= min.z && point.z <= max.z;
	}

}

```

**Reply by maluoi** [reply]
Physics will be removed in v0.4, but will remain in upcoming v0.3 builds. The only thing removed from this is the `Solid` class, all intersection and math primitives will remain in StereoKit's API!

If you want to cast rays against a Mesh or Model, try [Mesh.Intersect](https://stereokit.net/Pages/StereoKit/Mesh/Intersect.html) or [Model.Intersect](https://stereokit.net/Pages/StereoKit/Model/Intersect.html). There's also a [World.Raycast](https://stereokit.net/Pages/StereoKit/World/Raycast.html) that works well on HoloLens.

If you still need advanced physics simulations, like rigidbodies or joints, check out [Bepu Physics](https://github.com/bepu/bepuphysics2), and the [StereoKit sample](https://github.com/maluoi/StereoKitBepuPhysics) using Bepu.

**Comment by laultman** [comment]
@maluoi Thanks, I'll give the different intersection methods a try. In the case I have the intersection with the cylinder did not yield the result I needed. The intersection with the cylinder traverses the entire depth of the cylinder. I needed a precise location along the depth. I could have stacked thousands of cylinders with depth of the increment of my granular measurement. That introduces thousands of polygons and quickly exceeds the HoloLens capabilities. There had to be another way. Vec3.Dot method to the rescue.

I actually need the distance from the centerline of the cylinder (pipe) to a point outside the pipe and not in the same space (fingertip). I also needed to collide with an obstruction in the pipe (in the pipe space). I ended up shooting an imaginary bullet through the pipe and compared their vector's locations with some help with a matrix transform to get the hand and obstruction into the same space. Didn't need much fancy math and a single sqrt to get my distance.
